DOCSP-19059: logging framework (#276)
mongodb/docs-golang: source/fundamentals/logging.txt

.. _golang-logging:

=======
Logging
=======

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> Starting in version 1.12, the {+driver-short+} provides a logging

Author: switched it around so Go Driver is first

> Starting in version 1.12, the {+driver-short+} provides a logging

Author: How about we just omit the sentence altogether. It's only going to be on the v1.12 and later docs anyways.

> Starting in version 1.12, the {+driver-short+} provides a logging

Reviewer: love it

interface to record driver events. In this guide, you can learn how to
use the driver to configure logging for your application.

Enable Logging
--------------

To configure a logger on your ``Client`` instance, call the
``SetLoggerOptions()`` method when specifying your ``ClientOptions``

> object and pass these options to the ``Connect()`` method. The

Author: good call. Do you think the examples further down on this page need the `Connect()` call? I'm wondering if it's unneeded since the page is really about the `LoggerOptions` type.

``SetLoggerOptions()`` method takes a ``LoggerOptions`` type 
as a parameter, which contains the specifications that the driver uses
to configure the logger for your application.

The following code shows how to configure logging in a ``ClientOptions``
object and create a client with logging enabled:

.. code-block:: go
   :emphasize-lines: 4
   
   clientOptions := options.
   	Client().
   	ApplyURI(uri).
   	SetLoggerOptions(loggerOptions)

>    client, err := mongo.Connect(context.TODO(), clientOptions)

Author: I included code to make the most simple logger as possible.

Configure a Logger
------------------

To create a ``LoggerOptions`` object, call the ``options.Logger()``
method. The following table describes how to set properties on
``LoggerOptions`` to configure your logger. The first column contains
the ``LoggerOptions`` properties, the second column describes the
properties, and the third column describes the
setter methods:

.. list-table::
   :widths: 31 38 31
   :header-rows: 1

   * - Property
     - Description
     - Setter Method

   * - | ``ComponentLevels``

>        | **Type**: ``map[LogComponent]LogLevel``

Author: also added a newline between the setter methods and params

     - | A mapping of components, specified by ``LogComponent``, to log

>          severity levels, specified by ``LogLevel``. The driver uses the

Author: If they don't know what components are, they can read the relevant section that this table links to

         ``LogLevel`` for each ``LogComponent`` to determine if the log
         message should be generated. To learn more about
         ``LogComponent`` and ``LogLevel``, see :ref:`Logging Components
         and Severity Levels <golang-logging-component-severity>`.
     - | ``SetComponentLevel()``
       | **Parameters**: ``LogComponent``, ``LogLevel``

   * - | ``Sink``
       | **Type**: ``LogSink``
     - | The logging interface that the driver uses to log messages.
         The ``LogSink`` type is an interface that you can implement to
         provide a :ref:`custom sink <golang-logging-custom>` or integrate a :ref:`third-party
         logger <golang-logging-thirdparty>` for the driver's logs. If you don't set this
         property, driver uses the standard logging library.
     - | ``SetSink()``
       | **Parameter**: ``LogSink``

   * - | ``MaxDocumentLength``
       | **Type**: ``uint``

>        | **Default**: ``1000``

Author: `Sink` has a default but it's just that the driver uses a default logger. so it's not a value. And `ComponentLevels` is not optional, so no default

>      - | The maximum length in bytes of a document that the driver logs. If the

Author: I'll just change the wording. this describes the size of the log message that the driver prints

         underlying document is larger than this value, the driver
         truncates it and appends ellipses to the partial log message.
     - | ``SetMaxDocumentLength()``
       | **Parameter**: ``uint``

.. tip:: Write Logs to a Specific File
   
   By default, the standard logger emits messages to your console (``stderr``). You can
   specify a logging destination by setting the ``MONGODB_LOG_PATH``

>    environment variable to a filepath or ``stdout``.

Reviewer: S:

```suggestion
   environment variable to ``stdout`` or a filepath.
```

.. _golang-logging-component-severity:

Logging Components and Severity Levels
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can specify which components the driver will log against by

Reviewer: Q: Is logging "against" the right phrasing?

Author: This is the phrasing used in the [API documentation](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.12.0/mongo/options#LogComponent)

> You can specify which components the driver will log against by
> specifying the ``LogComponent`` type. The driver provides the following

Reviewer: I: Can we avoid "specify" twice?

```suggestion
You can use the ``LogComponent`` type to specify the
components the driver will log against. The driver provides the following
```

Author: reworded.

built-in specifications for ``LogComponent``:

- ``LogComponentAll``: enables logging for all components.
- ``LogComponentCommand``: enables command monitor logging.
- ``LogComponentTopology``: enables topology logging.
- ``LogComponentServerSelection``: enables server selection logging.

> - ``LogComponentConnection``: enables connection services logging.

Reviewer: Q: Would this be easier to read as a table? That has the added benefit of letting you include the corresponding enum values, which you mention in the next paragraph.

Ditto for LogLevel section later.

Author: Changed

You can also specify the logging component by using the enumeration of
the preceding specifications. The following code shows equivalent ways
of enabling command monitoring:

.. code-block:: go

   // Using named value
   comp := options.LogComponentCommand

   // Using enumeration

>    comp := options.LogComponent(1)

Reviewer: Would be clearer if using a table as mentioned above

> You can specify the log severity level by specifying the ``LogLevel``
> type. The driver provides the following

Reviewer: ```suggestion
To specify the log severity level, use the ``LogLevel``
type. The driver provides the following
```

built-in specifications for ``LogLevel``:

- ``LevelOff``: suppresses logging.
- ``LevelInfo``: enables logging of high-level informational messages.
- ``LevelDebug``: enables logging of detailed debug messages.

You can also specify the severity level by using the enumeration of
the preceding specifications. The following code shows equivalent ways
of enabling logging at the ``INFO`` level:

.. code-block:: go

   // Using named value
   lvl := options.LogLevelInfo

   // Using enumeration
   lvl := options.LogLevel(1)

.. important::

>    The {+driver-short+} currently supports only ``DEBUG`` level messages.

Reviewer: Q: Is "OFF" supported? I wonder whether we should include levels that the reader can't use.

Author: ``OFF`` is not really a logging level. I think it's important to describe all of the options or there could be confusion for users going back and forth between the API docs and these docs

Author: addressed and changed wording in the admonition

> To learn more about logging severity levels, see the Wikipedia entry on

Reviewer: S: Tip/Seealso admonition

the :wikipedia:`Syslog standard for message logging <Syslog#Severity_level>`.

Example
~~~~~~~

> This example shows how to configure a basic logger that uses the standard

Reviewer: Q: Does something specific make it "basic"? If not, I'd just drop it

Author: restructuring sentence

> logging library and prints log messages. The logger has the following

Reviewer: Q: Is there specific code that prints log messages here? If not, and it's standard behavior for a logger, I'd drop it

specifications:

> - The maximum document length is ``25``.

Reviewer: ```suggestion
- The maximum document length is ``25`` bytes.
```

- The logging component is ``LogComponentCommand``.
- The logging severity level is ``LogLevelDebug``.

.. literalinclude:: /includes/fundamentals/code-snippets/logging.go
   :language: go
   :dedent:
   :emphasize-lines: 1-4, 9
   :start-after: start-standard-logger
   :end-before: end-standard-logger

The following code performs an insert operation, which generates log
messages:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/logging.go
      :language: go
      :dedent:
      :start-after: start-insert
      :end-before: end-insert

>    .. output::

Reviewer: Q: Is this output what the user would see in the log?

Author: yes

      :language: console
      :visible: false

      {"command":"{\"insert\": \"testColl\",\"or...","commandName":"insert","databaseName":"testDB","driverConnectionId":1,"message":"Command started","operationId":0,"requestId":13,"serverConnectionId":97377,"serverHost":"...","serverPort":27017,"timestamp":...}
      {"commandName":"insert","driverConnectionId":1,"durationMS":19,"message":"Command succeeded","operationId":0,"reply":"{\"n\": {\"$numberInt\":\"1\"},...","requestId":13,"serverConnectionId":97377,"serverHost":"...","serverPort":27017,"timestamp":...}

.. _golang-logging-custom:

Use a Custom Logger
-------------------

If the standard logging library does not meet your needs, you can
implement a custom logger. By customizing your logging configuration,
you have more control over the content, format, and frequency of log messages.

To use a custom logger, define a logger struct and implement the
``Info()`` and ``Error()`` methods for the struct. Next, set the logger as
the ``LogSink`` for your ``Client`` by calling the ``SetSink()`` method on
your ``LoggerOptions`` instance.

Example
~~~~~~~

This example demonstrates how to define and implement a custom logger.

> 1. Define the ``CustomLogger`` struct.

Reviewer: A colon might tie the instruction to the code sample more closely:

```suggestion
1. Define the ``CustomLogger`` struct:
```

Ditto for other steps

   .. output::
      :language: console
      :visible: false

      level: 1 DEBUG, message: Connection pool created
      level: 1 DEBUG, message: Connection pool ready
      level: 1 DEBUG, message: Connection pool created
      level: 1 DEBUG, message: Connection pool ready
      level: 1 DEBUG, message: Connection pool created
      level: 1 DEBUG, message: Connection pool ready
      level: 1 DEBUG, message: Connection checkout started
      level: 1 DEBUG, message: Connection created
      level: 1 DEBUG, message: Connection ready
      level: 1 DEBUG, message: Connection checked out
      level: 1 DEBUG, message: Command started
      level: 1 DEBUG, message: Command succeeded
      level: 1 DEBUG, message: Connection checked in

.. _golang-logging-thirdparty:

Integrate Third-Party Loggers
-----------------------------

There are many third-party logging packages available in Go. To use
a third-party logger in your application, create a logger and assign it
as the sink in your ``LoggerOptions`` instance.

.. tip:: Logging Packages

   Some of the loggers you can integrate with the {+driver-short+}
   include :github:`logrus <sirupsen/logrus>`, :github:`zap

>    <uber-go/zap>`, and :github:`zerolog <rs/zerolog>`. For full examples
>    using each of these loggers, see the :github:`driver source code <mongodb/mongo-go-driver/tree/master/examples/_logger>`.

Reviewer: I: https://www.mongodb.com/docs/meta/style-guide/style/links-and-cross-references/

```suggestion
   <uber-go/zap>`, and :github:`zerolog <rs/zerolog>`. For full examples
   using each of these loggers, see the :github:`{+driver-short+} GitHub repository <mongodb/mongo-go-driver/tree/master/examples/_logger>`.
```

>    using each of these loggers, see the :github:`driver source code <mongodb/mongo-go-driver/tree/master/examples/_logger>`.

Reviewer: I: Feels more sensible after the following example. As written, a reader would think that they need to go to the repo to find an example.

Author: Moved and omitted logrus from the list as the page includes that example

Example
~~~~~~~

This example demonstrates how to define and integrate a ``logrus``

> logger into your application.

Reviewer: Q: Does this example differ meaningfully from the one on the repo? If not, I'd leave it out of the list.

Author: Yes, this shows how to implement a custom formatter (easy formatter)

1. Install the necessary packages.

Copy the following ``go get`` commands into your terminal to download

> the ``logrus`` packages needed for this example:

Reviewer: I: This formatting is funky; the paragraph kicks back outside the numbering, so it isn't clear how they're related.

S: Is there a different numbering option we could use here?

Author: I'm going to try to use the `step` formatting and see what that looks like.

> Copy the following ``go get`` commands into your terminal to download
> the ``logrus`` packages needed for this example:

Reviewer: ```suggestion
Run the following ``go get`` commands in your terminal to download
the ``logrus`` packages needed for this example:
```

   clientOptions := options.
   	Client().
   	ApplyURI(uri).
   	SetLoggerOptions(loggerOptions)

   client, err := mongo.Connect(context.TODO(), clientOptions)

Configure a Logger
------------------

To create a ``LoggerOptions`` object, call the ``options.Logger()``
method. The following table describes how to set properties on
a ``LoggerOptions`` type to configure your logger. The first column lists
the ``LoggerOptions`` properties, the second column describes the
properties, and the third column lists the corresponding setter method and
parameters for each property:

.. list-table::
   :widths: 31 38 31
   :header-rows: 1

   * - Property
     - Description
     - Setter Method

   * - | ``ComponentLevels``
       |
       | **Type**: ``map[LogComponent]LogLevel``
     - | A mapping of components to log severity levels. The driver uses the
         ``LogLevel`` for each ``LogComponent`` to determine if the log
         message should be generated.
       |
       | To learn more about ``LogComponent`` and ``LogLevel``, see
         :ref:`Logging Components and Severity Levels <golang-logging-component-severity>`.
     - | ``SetComponentLevel()``
       |

>        | **Parameters**: ``LogComponent``, ``LogLevel``

Author: This doesn't seem to be the convention we use [in other places](https://github.com/mongodb/docs-golang/blob/5da136c0089d5f3b8ddda927d57bb5518328f86a/source/fundamentals/transactions.txt#L63-L70), so I don't think we should start here.

   * - | ``Sink``
       |
       | **Type**: ``LogSink``
     - | The logging interface that the driver uses to log messages.
         The ``LogSink`` type is an interface that you can implement to
         provide a custom sink or integrate a third-party
         logger for the driver's logs. If you don't set this
         property, the driver uses the standard logging library.
       |
       | To learn more, see the :ref:`Use a Custom Logger
         <golang-logging-custom>` and :ref:`Integrate Third-Party
         Loggers <golang-logging-thirdparty>` sections of this guide.
     - | ``SetSink()``
       |
       | **Parameter**: ``LogSink``

   * - | ``MaxDocumentLength``
       |
       | **Type**: ``uint``
       | **Default**: ``1000``

>      - | The maximum length in bytes of log message that the driver emits. If the

Author: oops, yes

         message is larger than this value, the driver
         truncates it and appends ellipses to the partial log message.
     - | ``SetMaxDocumentLength()``
       |
       | **Parameter**: ``uint``

.. tip:: Write Logs to a Specific File
   
   By default, the standard logger emits messages to your console (``stderr``). You can
   specify a logging destination by setting the ``MONGODB_LOG_PATH``
   environment variable to ``stdout`` or a filepath.

.. _golang-logging-component-severity:

Logging Components and Severity Levels
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To specify the components that the driver logs against, set the
``LogComponent`` type. The following table describes built-in
specifications for ``LogComponent``:

.. list-table::
   :widths: 30 40 30
   :header-rows: 1

   * - Setting
     - Description
     - Enumeration

   * - ``LogComponentAll``
     - Enables logging for all components.
     - ``0``

   * - ``LogComponentCommand``

>      - Enables command monitor logging.

Author: removed all periods in these two tables

     - ``1``

   * - ``LogComponentTopology``
     - Enables topology logging.
     - ``2``

   * - ``LogComponentServerSelection``
     - Enables server selection logging.
     - ``3``

   * - ``LogComponentConnection``
     - Enables connection services logging.
     - ``4``

> You can specify the logging component by using either the setting name

Author: changed elsewhere too.

      :visible: false

      {"command":"{\"insert\": \"testColl\",\"or...","commandName":"insert","databaseName":"testDB","driverConnectionId":1,"message":"Command started","operationId":0,"requestId":13,"serverConnectionId":97377,"serverHost":"...","serverPort":27017,"timestamp":...}
      {"commandName":"insert","driverConnectionId":1,"durationMS":19,"message":"Command succeeded","operationId":0,"reply":"{\"n\": {\"$numberInt\":\"1\"},...","requestId":13,"serverConnectionId":97377,"serverHost":"...","serverPort":27017,"timestamp":...}

.. _golang-logging-custom:

Use a Custom Logger
-------------------

If the standard logging library does not meet your needs, you can
implement a custom logger. By customizing your logging configuration,
you have more control over the content, format, and frequency of log messages.

To use a custom logger, define a logger struct and implement the
``Info()`` and ``Error()`` methods for the struct. Next, set the logger as
the ``LogSink`` for your ``Client`` by calling the ``SetSink()`` method on
your ``LoggerOptions`` instance.

Example
~~~~~~~

This example demonstrates how to define and implement a custom logger.

.. procedure::

>    .. step:: Define the ``CustomLogger`` struct

Author: added periods to all steps

            :end-before: end-insert

         .. output::
            :language: console
            :visible: false

            level: 1 DEBUG, message: Connection pool created
            level: 1 DEBUG, message: Connection pool ready
            level: 1 DEBUG, message: Connection pool created
            level: 1 DEBUG, message: Connection pool ready
            level: 1 DEBUG, message: Connection pool created
            level: 1 DEBUG, message: Connection pool ready
            level: 1 DEBUG, message: Connection checkout started
            level: 1 DEBUG, message: Connection created
            level: 1 DEBUG, message: Connection ready
            level: 1 DEBUG, message: Connection checked out
            level: 1 DEBUG, message: Command started
            level: 1 DEBUG, message: Command succeeded
            level: 1 DEBUG, message: Connection checked in

.. _golang-logging-thirdparty:

Integrate Third-Party Loggers
-----------------------------

There are many third-party logging packages available in Go. To use
a third-party logger in your application, create a logger and assign it
as the sink in your ``LoggerOptions`` instance.

Example
~~~~~~~

> This example demonstrates how to define and integrate a :github:`logrus <sirupsen/logrus>`
> logger in your application.

Reviewer: We probably don't want to interrupt this example by sending them to GitHub

S: Remove link here and modify admonition below

```suggestion
This example demonstrates how  to integrate ``logrus``,
a third-party logging package, into your application.
```

.. procedure::

   .. step:: Install the necessary packages

      Run the following ``go get`` commands in your terminal to download
      the ``logrus`` packages needed for this example:
      
      .. code-block:: bash
      
         go get github.com/bombsimon/logrusr/v4
         go get github.com/sirupsen/logrus
         go get github.com/t-tomalak/logrus-easy-formatter

   .. step:: Define the ``logrus`` logger

>       The code creates a ``logrus`` logger with the following

Reviewer: ```suggestion
      The following code creates a ``logrus`` logger with the following
```

Author: changed so the sentence doesnt have two followings

      
      .. literalinclude:: /includes/fundamentals/code-snippets/logging.go
         :language: go
         :dedent:
         :emphasize-lines: 1, 3-6, 11
         :start-after: start-set-thirdparty-logger
         :end-before: end-set-thirdparty-logger

   .. step:: Perform operations

      The following code performs some CRUD operations, which generate log
      messages:

      .. io-code-block::
         :copyable: true

         .. input:: /includes/fundamentals/code-snippets/logging.go
            :language: go
            :dedent:
            :start-after: start-log-operations
            :end-before: end-log-operations

         .. output::
            :language: console
            :visible: false

            [DEBUG]: 2023-06-28 11:13:41 - Command started <insert>
            [DEBUG]: 2023-06-28 11:13:41 - Command succeeded <insert>
            [DEBUG]: 2023-06-28 11:13:41 - Command started <delete>
            [DEBUG]: 2023-06-28 11:13:41 - Command succeeded <delete>
            [DEBUG]: 2023-06-28 11:13:41 - Command started <update>
            [DEBUG]: 2023-06-28 11:13:41 - Command succeeded <update>

.. tip:: Logging Packages

>    Some of the other third-party loggers you can integrate with the {+driver-short+}
>    include :github:`zap <uber-go/zap>` and :github:`zerolog <rs/zerolog>`. For full examples
>    using these loggers, see the :github:`logging tests in the
>    {+driver-short+} Github repository <mongodb/mongo-go-driver/tree/master/examples/_logger>`.

Reviewer: ```suggestion
   You can find more information on third-party logging packages in their respective GitHub repositories:
   - :github:`logrus <sirupsen/logrus>`
   - :github:`zap <uber-go/zap>`
   - :github:`zerolog <rs/zerolog>`
   
   To see full code examples that integrate these loggers, see the :github:`logging tests in the
   {+driver-short+} Github repository <mongodb/mongo-go-driver/tree/master/examples/_logger>`.
```

      specifications:
      
      - The logger writes log messages to the console.

>       - The logger writes messages at the ``DebugLevel`` level.

Reviewer 2: Suggestion:
I think it could be good to stay consistent in describing the action the logger takes ("emits" / "logs" / "writes"). I would probably choose "logs" since it more precisely describes what a "logger" accomplishes.

   .. step:: Set the logger as the ``Sink`` for your ``Client``.

>       In this example, the logger logs against commands at the ``LevelDebug`` level.

Reviewer 2: Suggestions:

- I think "this example" is ambiguous because it is used to refer to the entirety of the section at the start. Instead, refer to the snippet.

- I think the explanation could be more accurate since it doesn't actually log anything in the code snippet, only sets it up.

- I think "against" should be omitted since "logs"sufficiently describes what the logger does to commands. Applies to all instances.

"In the following code example, the logger is configured to log commands at the LevelDebug level."
